src/main/java/com/krishagni/openspecimen/msk2/importer/CarsImporter.java

```diff
@@ -0,0 +1,278 @@
+				return req;
+			}
+		}
+		
+		return null;
+	}
+	
+	private SpecimenRequirementDetail createSr(List<SpecimenRequirementDetail> sprDetails) throws Exception {
+		SpecimenRequirementDetail sprDetail = sprDetails.iterator().next();
+		ResponseEvent<SpecimenRequirementDetail> resp = cpSvc.addSpecimenRequirement(request(sprDetail));
+		resp.throwErrorIfUnsuccessful();
+		
+		return resp.getPayload();
+	}
+
+	private SpecimenRequirementDetail updateSr(List<SpecimenRequirementDetail> srDetails, SpecimenRequirementDetail srFromDb) throws Exception {
+		SpecimenRequirementDetail sprDetail = srDetails.iterator().next();
+		sprDetail.setId(srFromDb.getId());
+		ResponseEvent<SpecimenRequirementDetail> resp = cpSvc.updateSpecimenRequirement(request(sprDetail));
+		resp.throwErrorIfUnsuccessful();
+		
+		return resp.getPayload();
+	}
+	
+	////////////////////////////////
+	//
+	// Collection Protocol Event
+	//
+	///////////////////////////////
+	
+	private void checkEvent(CollectionProtocolDetail cpDetail, CollectionProtocolEventDetail event) throws Exception {
+		CollectionProtocolEventDetail eventDetail = eventsMap.get(event.getEventLabel());
+		
+		if (eventDetail == null) {
+			CollectionProtocolEvent eventFromDb = getEventFromDb(event.getCpShortTitle(), event.getEventLabel());
+			eventDetail = eventFromDb != null ? CollectionProtocolEventDetail.from(eventFromDb) : null;
+		}
+		
+		if (eventDetail != null) {
+			updateEvent(cpDetail, eventDetail, event);
+		} else {
+			eventDetail = createEvent(cpDetail, event);
+		}
+		
+		eventsMap.put(event.getEventLabel(), eventDetail);
+	}
+
+	private CollectionProtocolEvent getEventFromDb(String cpShortTitle, String eventLabel) {
+		return daoFactory.getCollectionProtocolDao().getCpeByShortTitleAndEventLabel(cpShortTitle, eventLabel);
+	}
+	
+	private CollectionProtocolEventDetail createEvent(CollectionProtocolDetail cpDetail,
+			CollectionProtocolEventDetail event) throws Exception {
+		CollectionProtocolEventDetail eventDetail = new CollectionProtocolEventDetail();
+		eventDetail.setCpShortTitle(cpDetail.getShortTitle());
+		eventDetail.setEventLabel(event.getEventLabel());
+		
+		ResponseEvent<CollectionProtocolEventDetail> resp = cpSvc.addEvent(request(event));
+		resp.throwErrorIfUnsuccessful();
+		
+		return resp.getPayload();
+	}
+
+	private void updateEvent(CollectionProtocolDetail cpDetail, CollectionProtocolEventDetail eventDetail,
+			CollectionProtocolEventDetail event) {
+		
+	}
+	
+	//////////////////////////
+	//
+	// Collection Protocol
+	//
+	/////////////////////////
+	
+	private CollectionProtocolDetail checkCp(CollectionProtocolEventDetail event, CarsDetail object) throws Exception {
+		CollectionProtocolDetail cpDetail = new CollectionProtocolDetail();
+		cpDetail = cpMap.get(event.getCpShortTitle());
+		
+		if (cpDetail == null) {
+			cpDetail = getCpFromDB(event.getCpShortTitle());
+			if (cpDetail == null) {
+		        	cpDetail = createCp(object);
```

Comment:
It is not a good idea to have nested conditionals. It breaks the program linearity.

Instead, do this:

public CollectionProtocolDetail getCp(CollectionProtocolDetail input) {
  return cpMap.computeIfAbsent(input.getShortTitle(), () -> createIfAbsent(input));
}

private CollectionProtocolDetail createIfAbsent(CollectionProtocolDetail input) {
  CollectionProtocolDetail dbCp = getCpFromDb(input.getShortTitle());
  return dbCp != null ? dbCp : createCp(input);
}

Reply:
@vinayakapawar 
The createIfAbsent() method here throws an Exception (as getCpFromDb() method throws an Exception).
When I'm using cpMap.computeIfAbsent() I have to try-catch and handle the exception right there in getCp() method but we have to throw the exception to the upper method, as that one is @PlusTransactional. How do I propagate the Exception to upper method.
I searched for how to throw Exceptions from computeIfAbsent() method, throwing Exception from Lambda expressions, etc. There seems no way around it. 

I'm stumped, Can you please help?

```java
private CollectionProtocolDetail getCp(CollectionProtocolDetail input) throws Exception {
		return cpMap.computeIfAbsent(input.getShortTitle(), k -> {
			try {
				return createIfAbsent(input);
			} catch (Exception e) {
				// What to do after catching
			}
			return input;
		});
	}
```